utils_nlp/models/transformers/abstractive_summarization_bertsum.py

```diff
@@ -0,0 +1,815 @@
+            world_size (int, optional): Total number of GPUs that will be used. Defaults to 1.
+            rank (int): Rank of the current GPU. Defaults to -1.
+            validation_function
+        """
+
+        # get device
+        device, num_gpus = get_device(
+            num_gpus=num_gpus, gpu_ids=gpu_ids, local_rank=local_rank
+        )
+        # move model to devices
+        print("device is {}".format(device))
+        if checkpoint:
+            # the following line creats addtional processes on GPU 0
+            # point where memory use increase
+            checkpoint = torch.load(checkpoint, map_location="cpu")
+            self.model.load_checkpoint(checkpoint["model"])
+        self.model = move_model_to_device(model=self.model, device=device)
+        # init optimizer
+
+        # """
+        self.optim_bert = model_builder.build_optim_bert(
+            self.model,
+            optim=optimization_method,
+            visible_gpus=None,  # ",".join([str(i) for i in range(num_gpus)]), #"0,1,2,3",
+            lr_bert=learning_rate_bert,
+            warmup_steps_bert=warmup_steps_bert,
+            checkpoint=None,
+        )
+        self.optim_dec = model_builder.build_optim_dec(
+            self.model,
+            optim=optimization_method,
+            visible_gpus=None,  # ",".join([str(i) for i in range(num_gpus)]), #"0,1,2,3"
+            lr_dec=learning_rate_dec,
+            warmup_steps_dec=warmup_steps_dec,
+        )
+        # """
+
+        optimizers = [self.optim_bert, self.optim_dec]
+
+        self.amp = get_amp(fp16)
+        if self.amp:
+            self.model, optim = self.amp.initialize(
+                self.model, optimizers, opt_level=fp16_opt_level
+            )
+
+        global_step = 0
+        if checkpoint:
+            if checkpoint["optimizers"]:
+                for i in range(len(optimizers)):
+                    model_builder.load_optimizer_checkpoint(
+                        optimizers[i], checkpoint["optimizers"][i]
+                    )
+            if self.amp and "amp" in checkpoint and checkpoint["amp"]:
+                self.amp.load_state_dict(checkpoint["amp"])
+            if "global_step" in checkpoint and checkpoint["global_step"]:
+                global_step = checkpoint["global_step"] / world_size
+                print("global_step is {}".format(global_step))
+
+        self.model = parallelize_model(
+            model=self.model,
+            device=device,
+            num_gpus=num_gpus,
+            gpu_ids=gpu_ids,
+            local_rank=local_rank,
+        )
+
+        if local_rank == -1:
+            sampler = RandomSampler(train_dataset)
+        else:
+            sampler = DistributedSampler(
+                train_dataset, num_replicas=world_size, rank=rank
+            )
```

Comment:
I don't remember exactly, should rank=local_rank if you are doing multi-gpu DDP on single node?

Reply:
correct,but not the same when you use multiple nodes. What do you mean to say?

Reply:
nvm. I thought we could compute rank from local_rank to make the method easier to use, but we don't have enough information in this method. We have to calculate it outside.